### PR TITLE
Balancing and Fight restructuring

### LINKED ISSUE #158: Both i.hp <= 0
**Checklist**
I have:
- [*] Searched the Issues page for my problem/feature request
- [*] Updated all related software and Pokete it self

**Describe the bug**
IndexError when both pokete are at hp 0, possible due to effects such as burning and poison.
Caused by `winner = [i for i in players if i.hp > 0][0]`

**To Reproduce**
1. Run ./pokete.py
2. In a match, end with both pokete at hp 0

**Expected behavior**
The game crashes to the terminal with python error IndexError

**Environment**
 - OS: Artix 5.18.2
 - Pokete version 0.7.2
 - Python version 3.10.5
 - Scrap engine version 1.2.0

### PR DESCRIPTION
**Pull request by lxgr-linux**
- [x] Add xp scaling when fighting against higher level Poketes 
- [x] Add multiple Poketes to trainers
- [x] Add Fail chance to some attacks
- [x] Add second Poketes to some trainers
Fixes #158

This needs play testing.

### CODE REVIEW

**Comment by lxgr-linux** [comment]
@daynetran

**Comment by cognivore** [comment]
How does miss chance scale as a function of defense? Or is it an intrinsic thing amplified by "cry"?

**Comment by lxgr-linux** [comment]
Ready for review.

**Comment by lxgr-linux** [comment]
> How does miss chance scale as a function of defense? Or is it an intrinsic thing amplified by "cry"?

The chance of missing is influenced by the attacks and the Poketes miss chance. The Poketes misschance can be influenced by attacks like cry.

File: pokete_classes/fightmap.py [hunk 1]
```diff
@@ -197,226 +200,241 @@ def get_item(self, items, inv):
...
-            obj = [i for i in players if i != enem][-1]
-        for i in players:
+        index = providers.index(
+            sorted(
+                zip(
+                    [i.curr.initiative for i in providers],
+                    [1, 0],
+                    providers
+                )
+            )[-1][-1]
```
Comment (mustafaquraish) [inline pokete_classes/fightmap.py:308]:
```py
max(providers, key=lambda i: i.curr.initiative)
```

File: pokete_classes/fightmap.py [hunk 2]
```diff
@@ -197,226 +200,241 @@ def get_item(self, items, inv):
...
+        )
+        for i in [prov.curr for prov in providers]:
```
Comment (mustafaquraish) [inline pokete_classes/fightmap.py:310]: There's no need to create a copy of the list here, just do:
```py
for prov in providers:
    i = prov.curr
```

File: pokete_classes/fightmap.py [hunk 3]
```diff
@@ -197,226 +200,241 @@ def get_item(self, items, inv):
...
-                                                    if winner.player else ''))
-        if winner.player and info["type"] == "duel":
+            index += 1
+        # loser = [obj for obj in players if obj != winner][0]
+        _xp = sum(
+            [
+                poke.lose_xp + max(0, poke.lvl() - winner.curr.lvl())
+                for poke in loser.pokes
+            ]
+        ) * (2 if type(loser) is not NatureProvider else 1)
```
Comment (mustafaquraish) [inline pokete_classes/fightmap.py:374]: Remove the `[]` around the list comprehension, `sum()` can operator on generators AFAIK

File: pokete_classes/landscape.py [hunk 4]
```diff
@@ -26,18 +27,28 @@ def action(self, ob):
...
+                                weights=[i["rarity"] for _, i in pokes.items()]
+                            )[0],
+                            random.choices(
+                                list(
+                                    range(
+                                        self.arg_proto["minlvl"],
+                                        self.arg_proto["maxlvl"]
+                                    )
+                                )
+                            )[0],
```
Comment (mustafaquraish) [inline pokete_classes/landscape.py:46]: use `random.choice(...)`

File: pokete_classes/landscape.py [hunk 5]
```diff
@@ -26,18 +27,28 @@ def action(self, ob):
...
-                              0],
-                          player=False, shiny=(random.randint(0, 500) == 0)))
+            fm.fightmap.fight(
+                [
+                    self.figure,
+                    NatureProvider(
+                        Poke(
+                            random.choices(
+                                list(pokes),
+                                weights=[i["rarity"] for _, i in pokes.items()]
```
Comment (mustafaquraish) [inline pokete_classes/landscape.py:37]: No point looping over `.items()` and then ignoring the key, you're wasting tuple constructions
```py
weights=[i["rarity"] for i in pokes.values()]
```

File: pokete_classes/landscape.py [hunk 6]
```diff
@@ -26,18 +27,28 @@ def action(self, ob):
...
-                                                    self.arg_proto["maxlvl"])))[
-                              0],
-                          player=False, shiny=(random.randint(0, 500) == 0)))
+            fm.fightmap.fight(
+                [
+                    self.figure,
+                    NatureProvider(
+                        Poke(
+                            random.choices(
+                                list(pokes),
```
Comment (mustafaquraish) [inline pokete_classes/landscape.py:36]: Does this need to be a list? Can `random.choices()` not handle any iterable?

File: pokete_classes/fightmap.py [hunk 7]
```diff
@@ -61,79 +63,80 @@ def __init__(self, height, width):
...
-            obj.remove()
-        self.box.remove_c_obs()
-        for i in [player, enemy]:
-            for j in i.effects:
+        for prov in providers:
+            for obj in [
+                prov.curr.text_name, prov.curr.text_lvl, prov.curr.text_hp,
+                prov.curr.ico, prov.curr.hp_bar, prov.curr.tril, prov.curr.trir,
+                prov.curr.pball_small
+            ]:
```
Comment (mustafaquraish) [inline pokete_classes/fightmap.py:77]: Make this a tuple instead

File: pokete_classes/providers.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+    @property
+    def curr(self):
+        """Returns the currently used Pokete"""
+        return self.pokes[self.play_index]
+
+    def index_conf(self):
+        """Sets index correctly"""
+        self.play_index = self.pokes.index(
+            [poke for poke in self.pokes if poke.hp > 0][0]
```
Comment (mustafaquraish) [inline pokete_classes/providers.py:20]: You can avoid creating the entire filtered list with:
```py
next(poke for poke in self.pokes if poke.hp > 0)
```

File: pokete_classes/landscape.py (hunk 6 above)
Comment (lxgr-linux) [reply pokete_classes/landscape.py:36]: It only takes a list.

File: pokete_classes/fightmap.py [hunk 9]
```diff
@@ -197,226 +200,234 @@ def get_item(self, items, inv):
...
-                if (obj.identifier != "__fallback__"
-                        and not winner.player
-                        and any(p.hp > 0 for p in figure.pokes[:6])
-                        and ask_bool(self,
-                                     "Do you want to choose another Pokete?")):
-                    old_player = player
-                    players, player = self.choose_poke(figure, players,
-                                                       player, enemy)
-                    if old_player == player:
+                loser = providers[(providers.index(winner) + 1) % 2]
```
Comment (mustafaquraish) [inline pokete_classes/fightmap.py:331]: You already knew the index in the block above, so why not just save that instead? That will save you having to re-look through the whole list to find the index for the object.

File: pokete_classes/fightmap.py [hunk 10]
```diff
@@ -197,226 +200,234 @@ def get_item(self, items, inv):
...
+                if (
+                    loser.curr.player
+                    and any(p.hp > 0 for p in loser.pokes[:6])
+                    and ask_bool(self, "Do you want to choose another Pokete?")
+                ):
+                    success = self.choose_poke(loser)
+                    if not success:
                         break
+                elif (
+                    type(loser) is Trainer
```
Comment (mustafaquraish) [inline pokete_classes/fightmap.py:341]: `isinstance()`

File: pokete_classes/fightmap.py [hunk 11]
```diff
@@ -197,226 +200,234 @@ def get_item(self, items, inv):
...
-                             * (2 if info["type"] == "duel" else 1)
-        self.outp.outp(f"{winner.ext_name} won!" + (f'\nXP + {_xp}'
-                                                    if winner.player else ''))
-        if winner.player and info["type"] == "duel":
+            index += 1
+        # loser = [obj for obj in players if obj != winner][0]
+        _xp = sum(
+            poke.lose_xp + max(0, poke.lvl() - winner.curr.lvl())
+            for poke in loser.pokes
+        ) * (2 if type(loser) is not NatureProvider else 1)
```
Comment (mustafaquraish) [inline pokete_classes/fightmap.py:367]: `isinstance()`

File: pokete_classes/fightmap.py [hunk 12]
```diff
@@ -197,226 +200,234 @@ def get_item(self, items, inv):
...
+        self.outp.outp(
+            f"{winner.curr.ext_name} won!" +
+            (f'\nXP + {_xp}' if winner.curr.player else '')
+        )
+        if winner.curr.player and type(loser) is Trainer:
```
Comment (mustafaquraish) [inline pokete_classes/fightmap.py:372]: `isinstance()`

File: pokete_classes/fightmap.py [hunk 13]
```diff
@@ -428,105 +439,83 @@ def throw(self, obj, enem, info, chance, name):
             2: The win the game
             None: To let the enemy attack"""
 
-        if obj.identifier == "__fallback__" or info["type"] == "duel":
+        if type(enem) is not NatureProvider:
```
Comment (mustafaquraish) [inline pokete_classes/fightmap.py:442]: `not isinstance()`

**Review by mustafaquraish — APPROVED** [review]

**Pull request merged**